lib/media-controls-impl.js

```diff
@@ -0,0 +1,524 @@
+          break;
+      }
+      break;
+
+      case 'touchend':
+      case 'mouseup':
+        //
+        // If ending a long-press forward or backward, clear timer
+        //
+        if (this.intervalId) {
+           window.clearInterval(this.intervalId);
+           this.intervalId = null;
+        }
+        else if (this.dragging) {
+          // If ending a movement of the slider, end slider movement
+          this.handleSliderMoveEnd();
+        }
+
+        if (e.type === 'mouseup') {
+          // Don't listen for mousemove and mouseup until we get a mousedown
+          window.removeEventListener('mousemove', this, true);
+          window.removeEventListener('mouseup', this, true);
+          this.mouseEventHandlerRegistered = false;
+        }
+        break;
+
+    case 'loadedmetadata':
+      //
+      // Metadata has been loaded, now we can set the duration of the media
+      //
+      this.els.durationText.textContent = this.formatTime(this.mediaPlayer.duration);
+      break;
+
+    case 'play':
+      //
+      // Media is playing, display 'paused' icon
+      //
+      this.els.play.classList.remove('paused');
+
+      // Update l10n-id for the benefit of the screen reader
+      this.els.play.setAttribute('data-l10n-id', 'playbackPlay');
+      break;
+
+    case 'pause':
+      //
+      // Media is paused, display 'play' icon
+      //
+      this.els.play.classList.add('paused');
+
+      // Update l10n-id for the benefit of the screen reader
+      this.els.play.setAttribute('data-l10n-id', 'playbackPause');
+
+      // If the paused event comes when the media is at the end,
+      // set our 'played-till-end' flag. The one exception is when
+      // the paused event comes from pausing the media when the
+      // forward button was used to seek to the end while the media
+      // was playing. In this case, we don't consider the media being
+      // played until the end.
+      if (this.mediaPlayer.currentTime === this.mediaPlayer.duration) {
+        if (this.pausedAtEndWhilePlaying) {
+          this.pausedAtEndWhilePlaying = false;
+        }
+        else {
+          this.playedUntilEnd = true;
+        }
+      }
+      break;
+
+    case 'timeupdate':
+      //
+      // Update the progress bar and play head as the video plays
+      //
+      if (!this.mediaControlsElement.hidden) {
+        this.updateMediaControlSlider();
+      }
+      break;
+
+    case 'seeked':
+      //
+      // Update the position of the slider when the video position has been
+      // moved.
+      //
+      this.updateMediaControlSlider();
+      break;
+
+    case 'ended':
+      //
+      // Ignore ended events that occur while the user is dragging the slider
+      //
+      if (this.dragging) {
+        return;
+      }
+
+      // If the media was played until the end (as opposed to being forwarded
+      // to the end), position the player at the beginning of the video.
+      if (this.playedUntilEnd) {
```

Comment:
This is kind of complicated logic, and it is also app-specific. For the music app, when we reach the end we would want to move on to the next song, not go back to the start.  I wonder if the better solution would be to fire an event in the normal played-until-end case and let the app handle it.

For now, what if we just ignored ended events? And if you did that, can you get rid of the special case code on the paused event, too?